drivers/dp/swdp_bitbang.c

```diff
@@ -0,0 +1,632 @@
+/*
+ * Copyright (c) 2018-2019 PHYTEC Messtechnik GmbH
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file is based on SW_DP.c from DAPLink Interface Firmware.
+ * Copyright (c) 2009-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+/* Serial Wire Debug Port interface bit-bang driver */
+
+#define DT_DRV_COMPAT zephyr_swdp_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/swdp.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(swdp, CONFIG_DP_DRIVER_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+#define CPU_CLOCK		64000000U
+#else
+#define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+#define CLOCK_DELAY(swclk_freq, port_write_cycles) \
+	((CPU_CLOCK / 2 / swclk_freq) - port_write_cycles)
+
+/*
+ * Default SWCLK frequency in Hz.
+ * sw_clock can be used to overwrite this default value.
+ */
+#define SWDP_DEFAULT_SWCLK_FREQUENCY	1000000U
+
+#define DELAY_SLOW_CYCLES		3U
+
+struct sw_config {
+	struct gpio_dt_spec clk;
+	struct gpio_dt_spec dout;
+	struct gpio_dt_spec din;
+	struct gpio_dt_spec dnoe;
+	struct gpio_dt_spec noe;
+	struct gpio_dt_spec reset;
+	uint32_t port_write_cycles;
+	void *clk_reg;
+	void *dout_reg;
+	void *din_reg;
+	void *dnoe_reg;
+};
+
+struct sw_cfg_data {
+	uint32_t clock_delay;
+	uint8_t turnaround;
+	bool data_phase;
+	bool fast_clock;
+};
+
+static uint8_t sw_request_lut[16] = {0U};
+
+static void mk_sw_request_lut(void)
+{
+	uint32_t parity = 0U;
+
+	for (int request = 0; request < sizeof(sw_request_lut); request++) {
+		parity = request;
+		parity ^= parity >> 2;
+		parity ^= parity >> 1;
+
+		/*
+		 * Move A[3:3], RnW, APnDP bits to their position,
+		 * add start bit, stop bit(6), and park bit.
+		 */
+		sw_request_lut[request] =  BIT(7) | (request << 1) | BIT(0);
+		/* Add parity bit */
+		if (parity & 0x01U) {
+			sw_request_lut[request] |= BIT(5);
+		}
+	}
+
+	LOG_HEXDUMP_DBG(sw_request_lut, sizeof(sw_request_lut), "request lut");
+}
+
+static uint32_t ALWAYS_INLINE sw_get32bit_parity(uint32_t data)
+{
+	data ^= data >> 16;
+	data ^= data >> 8;
+	data ^= data >> 4;
+	data ^= data >> 2;
+	data ^= data >> 1;
+
+	return data & 1U;
+}
+
+static ALWAYS_INLINE void pin_delay_asm(uint32_t delay)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	__asm volatile ("movs r3, %[p]\n"
+			".start_%=:\n"
+			"subs r3, #1\n"
+			"bne .start_%=\n"
+			:
+			: [p] "r" (delay)
+			: "r3", "cc"
+			);
+#else
+#error "Not defined for this SoC family"
+#endif
+}
+
+static ALWAYS_INLINE void pin_platform_set(void *base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTSET = BIT(pin);
+#else
+#error "Not defined for this SoC family"
```

Comment:
Could you fallback to using the GPIO APIs automatically?

Reply:
GPIO API is too slow

Reply:
Is it so slow that it does not work at all? I mean if it's just slow it may be better than not working on at all, you could add a big #warning about it of course.

Also @cfriedt I remember seeing something about using native GPIO for bitbanging in one of your FPGA PRs as well, maybe we could have some optional common macros for that as part of the GPIO APIs?

Reply:
[ARM®DS-5 Version 5 ARM DSTREAM System and Interface Design Reference Guide Table 2-2 SWD timing requirements](https://developer.arm.com/documentation/100956/0529/Arm-DSTREAM-Target-Interface-Connections/Signal-descriptions/Serial-Wire-Debug?lang=en) specifies Tlow and Thigh min as 10 ns and max as 500 us. Note that the 500 us is the absolute maximum you have between SWCLK pulses including all overhead.

Note that the 500 us is the max you have from ARM itself, the minimum frequency required can be higher depending on target device. For example for [nRF51 Series](https://infocenter.nordicsemi.com/pdf/nRF51_RM_v3.0.pdf) the minimum SWDCLK is 125 kHz which translates to 8 us max (instead of 500 us).

Reply:
Wow, was about to answer something like "we can't do 2MHz with the standard API?!" but then I hacked an infinite `gpio_pin_toggle_dt()` on an nrf52dk and got a mere 860kHz. Well, thanks for explaining.

Reply:
Fallback to GPIO is now implemented.